Automattic/jetpack
# Premium Blocks: new implementation for Premium blocks

## Goal

This PR suggests a new implementation to handle premium blocks from a unique place when it's possible. Almost all code resides in the `paid-blocks/` folder with the expectation of `core/cover` block that has a dual-mode in terms of functionality based on the site plan.
It's a filter/hook approach which means we can set/add new premium blocks slightly in an easy way. On this iteration, we deal with `Jetpack`, `FSE`, and `core` blocks.
Also, it removes the Upgrade Nudge from all premium blocks.

### Improvements/Fixes list

* Show a banner for a Premium block in editor canvas
* Show a banner for a Premium block in Block Inspector
* Show an icon variation for Premium blocks.
* Avoid showing native alert modal when redirects to the checkout page
* Avoid overlapping banner with button toolbar (for core blocks, provably related with light wrapper feature)

The design and requirements are defined (partially) in the following issues list:

* (8P) Premium Blocks: Add Upgrade banner to Placeholder [#44056](https://github.com/Automattic/wp-calypso/issues/44056)

Fixes https://github.com/Automattic/wp-calypso/issues/44056
Link https://github.com/Automattic/wp-calypso/issues/44057
Fixes https://github.com/Automattic/wp-calypso/issues/44048
Fixes https://github.com/Automattic/wp-calypso/issues/44168

## What are Premium blocks?

All blocks that require a paid plan to increase its functionality.

### Current Premium Blocks list

- Simple Payments (Jetpack)
- Recurring Payments (Jetpack)
- Premium Content (FSE)
- Donations (FSE/disabled so far)
- Calendly (Jetpack)
- OpenTable (Jetpack)
- Video (core)
- Audio (core)
- Cover (core/partially)
- WhatsApp (Jetpack)

## Design & Behaviour 

### Block Icon

The first sign for the user that the block is premium is its Icon. All premium blocks show a little red circle ⭕️ in their icon.

Inline Inserter | Menu Inserter | 
--------------|---------------
<img src="https://user-images.githubusercontent.com/77539/88173270-ab085780-cbf8-11ea-8acc-91c5f336e3c8.png" width="300px" /> | <img width="360" alt="Screen Shot 2020-07-22 at 8 52 47 AM" src="https://user-images.githubusercontent.com/77539/88173356-cc694380-cbf8-11ea-8548-4d3eef965033.png">
Block Iinspector | BlockToolbar 
<img width="285" alt="Screen Shot 2020-07-22 at 9 18 55 AM" src="https://user-images.githubusercontent.com/77539/88175603-739baa00-cbfc-11ea-94c6-19adfa9301a9.png"> | <img width="298" alt="Screen Shot 2020-07-22 at 9 19 55 AM" src="https://user-images.githubusercontent.com/77539/88175926-ead13e00-cbfc-11ea-8f24-f27a5b228b7b.png">

 ### Block Banner

When a premium block is inserted in the editor, it immediately shows the upgrade banner above it (with the exception of `core/cover` block). When the block is not selected, the banner disappears. Also, the banner is shown in the block-inspector/sidebar of the editor.

Editor Canvas | Block Inspector
--------------|---------------
<img src="https://user-images.githubusercontent.com/77539/88176458-b01bd580-cbfd-11ea-92e7-6f625ae5b7dc.png" width="600px" /> | <img src="https://user-images.githubusercontent.com/77539/88178403-b790ae00-cc00-11ea-885a-f02b1c54bb57.png" width="300px" />


#### Changes proposed in this Pull Request:


* Extend all blocks that should be treated as premium blocks.

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



Maybe.

## Testing instructions:

### Testing in dotcom

* Apply these changes to your SB (D46454-code)
* Sandbox a dotcom Simple site
* Go to post edit
* Add a paid block. Use the `premium` term in the blocks selector to get all premium blocks. Confirm you see the premium icon for these blocks.

<img src="https://user-images.githubusercontent.com/77539/88184307-e90d7780-cc08-11ea-961f-f890c80a7780.png" width="300px" />


* Confirm you see the new banner to upgrade the plan, as well as the premium icon.

<img src="https://user-images.githubusercontent.com/77539/88184449-19edac80-cc09-11ea-9955-a6c7ccba56cb.png" width="500px" />

* Confirm that clicking on the Upgrade button, it brings you to the checkout page.

### Saving post

This PR also handles saving post when the user clicks on the Upgrade button. Check that:

* If the post doesn't have changes, clicking on the Upgrade button leads to the checkout page immediately. 

No changes:
<img src="https://user-images.githubusercontent.com/77539/87611690-0c9f6200-c6df-11ea-944b-72b83d1a43b2.png" width="300px" />

* If there are unsaved changes, confirm that when the user clicks on the Upgrade button, the app saves the post before to perform the redirect. Confirm that it doesn't show the alert modal:

<img src="https://user-images.githubusercontent.com/77539/88185573-77cec400-cc0a-11ea-8d86-1a54b706b5ef.png" width="400px" />

Unsaved changes:
<img src="https://user-images.githubusercontent.com/77539/87611843-728be980-c6df-11ea-8d6a-afd2899f195a.png" width="300px" />

After clicking on the Upgrade button:

<img src="https://user-images.githubusercontent.com/77539/87611935-9e0ed400-c6df-11ea-8f1a-b1201d617c92.png" width="300px" />

In general, confirm that the Upgrade Nudge is not shown anymore

<img src="https://user-images.githubusercontent.com/77539/88185432-4fdf6080-cc0a-11ea-96c3-8704416deb44.png" width="500px" />

#### Proposed changelog entry for your changes:

* Add upgrade plan banner to blocks that require a paid plan.
